engine/chainstate/chainstate.go

```diff
@@ -36,34 +35,31 @@ func (c *Client) SupportedBroadcastFormats() HexFormatFlag {
 	}
 }
 
-// Broadcast will attempt to broadcast a transaction using the given providers
-func (c *Client) Broadcast(ctx context.Context, id, txHex string, format HexFormatFlag, timeout time.Duration) (string, error) {
-	// Basic validation
-	if len(id) < 50 {
-		return "", ErrInvalidTransactionID
-	} else if len(txHex) <= 0 { // todo: validate the tx hex
-		return "", ErrInvalidTransactionHex
-	}
-
-	// Debug the id and hex
-	c.DebugLog("tx_id: " + id)
-	c.DebugLog("tx_hex: " + txHex)
-
-	// Broadcast or die
-	successCompleteCh := make(chan string)
-	errorCh := make(chan string)
-
-	go c.broadcast(ctx, id, txHex, format, timeout, successCompleteCh, errorCh)
+type BroadcastResult struct {
+	Provider string
+	Failure  *BroadcastFailure
+}
+type BroadcastFailure struct {
+	InvalidTx bool
+	Error     error
+}
 
-	// wait for first success
-	success := <-successCompleteCh
-	if success != "" {
-		return success, nil
+// Broadcast will attempt to broadcast a transaction using the given providers
+func (c *Client) Broadcast(ctx context.Context, id, txHex string, format HexFormatFlag, timeout time.Duration) *BroadcastResult {
+	results := make(chan *BroadcastResult)
+	go c.broadcast(ctx, id, txHex, format, timeout, results)
+
+	failures := make([]*BroadcastResult, 0)
+
+	for r := range results {
+		if r.Failure != nil {
+			failures = append(failures, r)
+		} else {
+			return r // wait for first success
```

Comment:
I assume, that we don't want to do anything with the `failures` slice which can actually be not-empty here.

Reply:
I added better comment in code